### DISCUSSION TITLE
Question regarding StringContainerInterface with XML output

### DISCUSSION THREAD

**Discussion by tao** [thephpleague/commonmark#999] [Q&A]
I have been using an extension [commonmark-hint-extension](https://github.com/ueberdosis/commonmark-hint-extension) and run into a little bug or question regarding StringContainerInterface.

The class `Hint` extends StringContainerInterface like this:

```
// Hint.php
<?php

namespace Ueberdosis\CommonMark;

use League\CommonMark\Node\Block\AbstractBlock;
use League\CommonMark\Node\StringContainerInterface;

class Hint extends AbstractBlock implements StringContainerInterface
{
```

I am using the example from the plugin README file:

```markdown
:::important Warning!
This is how the **Markdown** looks.
:::
```

However when I render this as XML it outputs the text twice:

```xml
"""
  <?xml version="1.0" encoding="UTF-8"?>
  <document xmlns="http://commonmark.org/xml/1.0">
      <hint_block heading="Warning!" type="important">This is how the **Markdown** looks.
  
          <text>This is how the </text>
          <strong>
              <text>Markdown</text>
          </strong>
          <text> looks.</text>
      </hint_block>
  </document>
  """
```

You can see the text `This is how the **Markdown** looks` exists in plain text and then duplicated in XML text tags. When I remove this interface then it works fine:

```xml
"""
  <?xml version="1.0" encoding="UTF-8"?>
  <document xmlns="http://commonmark.org/xml/1.0">
      <hint_block heading="Warning!" type="important">
          <text>This is how the </text>
          <strong>
              <text>Markdown</text>
          </strong>
          <text> looks.</text>
      </hint_block>
  </document>
  """
```

I don't really understand the purpose of this interface and if it is required and working as expected or if this is a bug in Commonmark or if this is a bug in the package for implementing StringContainerInterface...

Any advice?

**Comment by colinodell** [comment]
`StringContainerInterface` should only be applied to leaf nodes - things like `Text` or `HtmlBlock` or `FencedCode` nodes that ONLY contain literal text instead of child nodes.

It looks like this extension is not following that behavior. In fact, `HintParser` is both [storing literal text](https://github.com/ueberdosis/commonmark-hint-extension/blob/bc258cf3fd382dcb7cc4b5f5a8b96a6f0beef838/src/HintParser.php#L84) AND [parsing text into child inline nodes](https://github.com/ueberdosis/commonmark-hint-extension/blob/bc258cf3fd382dcb7cc4b5f5a8b96a6f0beef838/src/HintParser.php#L51).  It should not be doing both of these things.  This is why you see the text to appear in the rendered XML as both the literal text and those child nodes.
